### ISSUE TITLE
SVG graphic could not be built

### ISSUE THREAD

**Issue by Intelligent2013** [metanorma/mn2pdf#26]
New issue 'SVG graphic could not be built' (https://github.com/metanorma/mn-native-pdf/issues/148):
```
SEVERE: SVG graphic could not be built. Reason: org.apache.batik.bridge.BridgeException: Content is not allowed in prolog.
org.apache.batik.bridge.BridgeException: Content is not allowed in prolog.
	at org.apache.batik.bridge.BridgeContext.getReferencedNode(BridgeContext.java:788)
	at org.apache.batik.bridge.BridgeContext.getReferencedElement(BridgeContext.java:804)
	at org.apache.batik.bridge.CSSUtilities.convertClipPath(CSSUtilities.java:718)
	at org.apache.batik.bridge.AbstractGraphicsNodeBridge.buildGraphicsNode(AbstractGraphicsNodeBridge.java:146)
	at org.apache.batik.bridge.GVTBuilder.buildGraphicsNode(GVTBuilder.java:224)
	at org.apache.batik.bridge.GVTBuilder.buildComposite(GVTBuilder.java:171)
	at org.apache.batik.bridge.GVTBuilder.buildGraphicsNode(GVTBuilder.java:219)
	at org.apache.batik.bridge.GVTBuilder.buildComposite(GVTBuilder.java:171)
	at org.apache.batik.bridge.GVTBuilder.buildGraphicsNode(GVTBuilder.java:219)
	at org.apache.batik.bridge.GVTBuilder.buildComposite(GVTBuilder.java:171)
	at org.apache.batik.bridge.GVTBuilder.buildGraphicsNode(GVTBuilder.java:219)
	at org.apache.batik.bridge.GVTBuilder.buildComposite(GVTBuilder.java:171)
	at org.apache.batik.bridge.GVTBuilder.build(GVTBuilder.java:82)
	at org.apache.fop.render.pdf.PDFImageHandlerSVG.handleImage(PDFImageHandlerSVG.java:104)
	at org.apache.fop.render.intermediate.AbstractIFPainter.drawImage(AbstractIFPainter.java:249)
	at org.apache.fop.render.intermediate.AbstractIFPainter.drawImage(AbstractIFPainter.java:205)

```

https://issues.apache.org/jira/browse/BATIK-1227 (Status: Open)
https://issues.apache.org/jira/browse/BATIK-1090 (Status: Open)

**Comment by Intelligent2013** [comment]
Additional information - some SVG images display ok.

**Comment by ronaldtse** [comment]
From #148 @Intelligent2013 

Regarding issue "SEVERE: SVG graphic could not be built. Reason: org.apache.batik.bridge.BridgeException: Content is not allowed in prolog.":

https://issues.apache.org/jira/browse/BATIK-1227 (Status: Open)
https://issues.apache.org/jira/browse/BATIK-1090 (Status: Open)

I've turned off (on my laptop) PDF/UA feature and found that image (in `<figure id="fig-surface-3d-or-volume-2d-elements-triangle-cubic">`) contains some text with font Symbol:
```
<text
           id="text32"
           style="font-variant:normal;font-weight:normal;font-size:12px;font-family:Symbol;-inkscape-font-specification:Symbol;writing-mode:lr-tb;fill:#000000;fill-opacity:1;fill-rule:nonzero;stroke:none"
           transform="matrix(1,0,0,-1,115.28,164.92)">...
```

SVG renders  in PDF as vector images with text (not bitmap), therefore Symbol font (or another fonts) should be determined in FOP config or replaced to another.

@ronaldtse In the current version of mn2pdf a few SVG show in PDF - with font-face issue. But there are a lot of SVG, which batik can't process. Looks like these SVG contain 'url' function that batik can't process. This error you should see:
```
SEVERE: SVG graphic could not be built. Reason: org.apache.batik.bridge.BridgeException: Content is not allowed in prolog.
```

**Comment by Intelligent2013** [comment]
Regarding issue:
```
SEVERE: SVG graphic could not be built. Reason: org.apache.batik.bridge.BridgeException: Content is not allowed in prolog.
org.apache.batik.bridge.BridgeException: Content is not allowed in prolog.
```
I've made some experiments, actions and tests. Preliminary result is here:
- first, I get first svg from source xml and converted it from base64 into svg xml file
- to make sure that patch https://issues.apache.org/jira/browse/BATIK-1227 fixing the issue, I've made batik-util jar with updated class `ParsedURLDataProtocolHandler` and replaced original jar with this new in the folder with batik-rasterizer utility (to convert svg graphics into raster image (png)) and it's working properly (without patch it interrupts with same error 'Content is not allowed in prolog')
- then put ParsedURLDataProtocolHandler.class file into mn2pdf.jar, and run it - there isn't  error: `'SEVERE: SVG graphic could not be built. Reason: org.apache.batik.bridge.BridgeException: Content is not allowed in prolog'`, but there is a new error `OutOfMemoryError: Java heap space`                                                             :
```
Exception in thread "main" java.lang.OutOfMemoryError: Java heap space                                                                              at java.base/java.util.Arrays.copyOf(Unknown Source)                                                                                        at java.base/java.lang.AbstractStringBuilder.ensureCapacityInternal(Unknown Source)                                                         
at java.base/java.lang.AbstractStringBuilder.append(Unknown Source)
at java.base/java.lang.StringBuilder.append(Unknown Source)
at org.apache.batik.util.ParsedURLDataProtocolHandler$DataParsedURLData.toString(ParsedURLDataProtocolHandler.java:151)
at org.apache.batik.util.ParsedURL.toString(ParsedURL.java:246)
at org.apache.batik.css.engine.value.AbstractValueFactory.resolveURI(AbstractValueFactory.java:41)                                          
at org.apache.batik.css.engine.value.svg.MaskManager.createValue(MaskManager.java:94)
at org.apache.batik.css.engine.CSSEngine$1.property(CSSEngine.java:744)
at org.apache.batik.css.engine.CSSEngine.getCascadedStyleMap(CSSEngine.java:767)
at org.apache.batik.css.engine.CSSEngine.getComputedStyle(CSSEngine.java:868)
at org.apache.batik.bridge.CSSUtilities.getComputedStyle(CSSUtilities.java:81)
at org.apache.batik.bridge.CSSUtilities.convertDisplay(CSSUtilities.java:563)
at org.apache.batik.bridge.GVTBuilder.buildGraphicsNode(GVTBuilder.java:206)
at org.apache.batik.bridge.GVTBuilder.buildComposite(GVTBuilder.java:171)                                      
```
- tried to increase Java heap space by `-Xmx1024m`, `-Xmx2048m`, `-Xmx4096m`, `-Xmx6000m` - again java.lang.OutOfMemoryError exception
- tried to decode base64 svg into svg xml file on disk and replaced `image/@src` to path into local file - and it's working! even with 1024m heap space.
- tried decode base64 and move it from `image/@src` inside `image` tag, and remove header `<?xml version="1.0" encoding="UTF-8" standalone="no"?>` (otherwise source xml doesn't correct) - and it's working also.

This tests were for only one svg item, but real xml source has 300 svg images.
The preliminary result: in mn2pdf (with batik, recommended by fop) we need to encode svg images from base64 to external temp folder on disk before PDF processing.

> Otherwise, can we use another SVG library like JFree? (https://github.com/jfree/jfreesvg)
I'll try this library, may be it will work without preliminary svg deserialization.

**Comment by Intelligent2013** [comment]
I can't find how to integrate jreesvg into fop and found here http://www.object-refinery.com/blog/blog-20140423.html that "_Apache Batik provides a *lot* of other SVG-related functions that are outside the scope of JFreeSVG_".
Therefore I've updated mn2pdf for extract svg base64 data into temporary folder in svg xml format, and then using links to local disk to these files.
Now the document iso-10303-104 process successfully.

Updated version here: https://github.com/metanorma/mn2pdf/releases/download/v1.14/mn2pdf-1.14.jar
